File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/opmodes/tuning/FlywheelsTuningBase.java
package org.firstinspires.ftc.teamcode.opmodes.tuning;

import com.acmerobotics.dashboard.config.Config;
import com.qualcomm.robotcore.hardware.DcMotorEx;
import com.qualcomm.robotcore.hardware.DcMotorSimple;

import org.firstinspires.ftc.teamcode.opmodes.base.MarsLinearOpMode;

@Config
public abstract class FlywheelsTuningBase extends MarsLinearOpMode {
    public static class HardwareConfig {
        public boolean MOTORS_COUPLED = true;
        public String[] MOTOR_NAMES = {"flywheel", "flywheel2"};
        public DcMotorSimple.Direction[] MOTOR_DIRECTIONS = {
            DcMotorSimple.Direction.REVERSE, DcMotorSimple.Direction.REVERSE
        };
    }

    public static HardwareConfig HARDWARE = new HardwareConfig();

    protected DcMotorEx[] motors;

    protected void initHardware() {
        initRobot();

        motors = new DcMotorEx[HARDWARE.MOTOR_NAMES.length];
        for (int i = 0; i < HARDWARE.MOTOR_NAMES.length; i++) {
            motors[i] = hardwareMap.get(DcMotorEx.class, HARDWARE.MOTOR_NAMES[i]);
            motors[i].setMode(DcMotorEx.RunMode.RUN_WITHOUT_ENCODER);
            motors[i].setDirection(HARDWARE.MOTOR_DIRECTIONS[i]);
            motors[i].setZeroPowerBehavior(DcMotorEx.ZeroPowerBehavior.FLOAT);
        }
    }

    public void stopMotors() {
        for (DcMotorEx m : motors) m.setPower(0);
    }

    @Override
    public final void runOpMode() throws InterruptedException {
        initHardware();
        runOpModeInternal();
    }

    protected abstract void runOpModeInternal() throws InterruptedException;
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/opmodes/tuning/FlywheelsTuningBase.kt
package org.firstinspires.ftc.teamcode.opmodes.tuning

import com.qualcomm.robotcore.hardware.DcMotor

import com.acmerobotics.dashboard.config.Config
import com.qualcomm.robotcore.hardware.DcMotorEx
import com.qualcomm.robotcore.hardware.DcMotorSimple
import org.firstinspires.ftc.teamcode.opmodes.base.MarsLinearOpMode

@Config
abstract class FlywheelsTuningBase : MarsLinearOpMode() {
    class HardwareConfig {
        @JvmField var MOTORS_COUPLED = true
        @JvmField var MOTOR_NAMES = arrayOf("flywheel", "flywheel2")
        @JvmField
        var MOTOR_DIRECTIONS = arrayOf(
            DcMotorSimple.Direction.REVERSE,
            DcMotorSimple.Direction.REVERSE,
        )
    }

    companion object {
        @JvmField
        var HARDWARE = HardwareConfig()
    }

    protected lateinit var motors: Array<DcMotorEx>

    protected fun initHardware() {
        initRobot()

        motors = Array(HARDWARE.MOTOR_NAMES.size) { i ->
            hardwareMap.get(DcMotorEx::class.java, HARDWARE.MOTOR_NAMES[i]).also { motor ->
                motor.mode = DcMotor.RunMode.RUN_WITHOUT_ENCODER
                motor.direction = HARDWARE.MOTOR_DIRECTIONS[i]
                motor.zeroPowerBehavior = DcMotor.ZeroPowerBehavior.FLOAT
            }
        }
    }

    fun stopMotors() {
        for (m in motors) m.power = 0.0
    }

    final override fun runOpMode() {
        initHardware()
        runOpModeInternal()
    }

    protected abstract fun runOpModeInternal()
}
